### PR TITLE
Use a sparse encoding for metadata tables

### PR DESCRIPTION
**Pull request by cjgillot**
r? @ghost

### CODE REVIEW

File: compiler/rustc_metadata/src/rmeta/table.rs [hunk 1]
```diff
@@ -16,76 +16,34 @@ use tracing::debug;
...
-
-    /// Read a `Self` value (encoded as `Self::BYTE_LEN` bytes),
-    /// from `&b[i * Self::BYTE_LEN..]`, returning `None` if `i`
-    /// is not in bounds, or `Some(Self::from_bytes(...))` otherwise.
-    fn maybe_read_from_bytes_at(b: &[u8], i: usize) -> Option<Self>;
-    /// Write a `Self` value (encoded as `Self::BYTE_LEN` bytes),
-    /// at `&mut b[i * Self::BYTE_LEN..]`, using `Self::write_to_bytes`.
-    fn write_to_bytes_at(self, b: &mut [u8], i: usize);
-}
+    /// This should be `[u8; BYTE_LEN]`;
```
Comment (fbstj) [inline compiler/rustc_metadata/src/rmeta/table.rs:20]: what `BYTE_LEN` is this referring to now? is it still an understood quantity in this context even though it's removed from this trait?

File: compiler/rustc_metadata/src/rmeta/table.rs [hunk 2]
```diff
@@ -250,62 +236,72 @@ where
...
     }
 }
 
 impl<I: Idx, T> LazyMeta for Table<I, T>
 where
-    Option<T>: FixedSizeEncoding,
+    T: FixedSizeEncoding,
 {
-    /// Number of bytes in the data stream.
-    type Meta = usize;
```
Comment (cjgillot) [inline compiler/rustc_metadata/src/rmeta/table.rs:303]: TODO: restore a comment here.

File: compiler/rustc_metadata/src/rmeta/table.rs [hunk 3]
```diff
@@ -250,62 +236,72 @@ where
...
     pub(crate) fn set<const N: usize>(&mut self, i: I, value: T)
     where
-        Option<T>: FixedSizeEncoding<ByteArray = [u8; N]>,
+        T: FixedSizeEncoding<ByteArray = [u8; N]>,
     {
         // FIXME(eddyb) investigate more compact encodings for sparse tables.
         // On the PR @michaelwoerister mentioned:
         // > Space requirements could perhaps be optimized by using the HAMT `popcnt`
         // > trick (i.e. divide things into buckets of 32 or 64 items and then
         // > store bit-masks of which item in each bucket is actually serialized).
```
Comment (cjgillot) [inline compiler/rustc_metadata/src/rmeta/table.rs:266]: The comment should be dropped.

**Comment by bjorn3** [comment]
For html5ever-0.26.0-debug this seems to be a 6% decrease in metadata size.

**Pull request closed without merging**